File: main/course-schedule-ii/course-schedule-ii-scratch.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public int[] findOrder(final int numCourses, final int[][] prerequisites) {
        final List<List<Integer>> adj = buildAdjacencyList(numCourses,
                                                           prerequisites);

        return reverseTopologicalSort(adj).stream().mapToInt(x -> x).toArray();
    }

    private static List<List<Integer>>
    buildAdjacencyList(int vertexCount, final int[][] edges) {
        // create an empty adjacency list with the correct number of rows
        final List<List<Integer>> adj = new ArrayList<>(vertexCount);
        while (vertexCount-- != 0) adj.add(new ArrayList<>());

        // populate the adjacency list with information about all the edges
        for (final int[] edge : edges) adj.get(edge[0]).add(edge[1]);
        return adj;
    }

    private static List<Integer>
    reverseTopologicalSort(final List<List<Integer>> adj) {
        final Color[] vis = new Color[adj.size()];
        Arrays.fill(vis, Color.WHITE);

        final List<Integer> out = new ArrayList<>(adj.size());

        for (int src = 0; src != adj.size(); ++src) {
            if (hasCycleFrom(adj, vis, out, src)) {
                out.clear();
                break;
            }
        }

        return out;
    }

    private static boolean
    hasCycleFrom(final List<List<Integer>> adj, final Color[] vis,
                 final List<Integer> out, final int src) {
        switch (vis[src]) {
        case WHITE:
            vis[src] = Color.GRAY;

            for (final int dest : adj.get(src))
                if (hasCycleFrom(adj, vis, out, dest)) return true;

            vis[src] = Color.BLACK;
            out.add(src);
            return false;

        case GRAY:
            return true;

        case BLACK:
            return false;
        }

        throw new AssertionError("invalid visitation state");
    }
}

/** Visitation states for traversing a directed graph. */
enum Color {
    /** Unvisited */
    WHITE,

    /** Visited but not yet fully explored */
    GRAY,

    /** Visited and fully explored */
    BLACK
}

final class UnitTest {
    public static void main(final String[] args) {
        final int[][] a = { {1, 0} };
        test(2, a);

        final int[][] b = { {1, 0}, {2, 0}, {3, 1}, {3, 2} };
        test(4, b);

        final int[][] c = { {2, 1}, {1, 0}, {2, 0}, {3, 1}, {3, 2} };
        test(4, c);

        final int[][] d = { {1, 2}, {1, 0}, {2, 0}, {3, 1}, {3, 2} };
        test(4, d);

        final int[][] e = { { 0, 1 }, { 1, 0 } };
        test(2, e);

        final int[][] f = { };
        test(7, f);

        final int[][] g = { {2, 4}, {1, 5} };
        test(7, g);
    }

    private static void
    test(final int numCourses, final int[][] prerequisites) {
        final int[] order = new Solution().findOrder(numCourses, prerequisites);
        System.out.println(Arrays.toString(order));
    }
}

File: main/course-schedule-ii/course-schedule-ii-scratch.cpp
#include <algorithm>
#include <cassert>
#include <functional>
#include <vector>

using namespace std;

class Solution {
public:
    static vector<int> findOrder(int numCourses,
                                 const vector<pair<int, int>>& prerequisites);
};

namespace {
    vector<vector<int>>
    build_adjacency_list(const int vertex_count,
                         const vector<pair<int, int>>& edges)
    {
        vector<vector<int>> adj (vertex_count);
        
        for (const auto edge : edges) {
            assert(0 <= edge.first && edge.first < vertex_count);
            assert(0 <= edge.second && edge.second < vertex_count);
            adj[edge.first].push_back(edge.second);
        }
        
        return adj;
    }
    
    // Visitation states for directed graph traversal.
    enum class Color {
        white,  // unvisited
        gray,   // visited but not fully explored
        black   // visited and fully explored
    };
    
    // Performs reverse topological sort by depth-first traversal.
    // When the graph is cyclic, returns an empty vector.
    vector<int> reverse_topological_sort(const vector<vector<int>>& adj)
    {
        vector<int> out;
        out.reserve(adj.size());
        
        vector<Color> vis (adj.size());
        
        // Tells if a directed graph has a cycle. For as long as no cycle is
        // found, also builds a reverse topological sort as it retreats.
        const function<bool(int)> has_cycle_from = [&](const int cur) {
            switch (vis[cur]) {
            case Color::white:
                {
                    vis[cur] = Color::gray;
                    
                    const auto& row = adj[cur];
                    if (any_of(cbegin(row), cend(row), has_cycle_from))
                        return true;
                    
                    vis[cur] = Color::black;
                    out.push_back(cur);
                    return false;
                }

            case Color::gray:
                return true;

            case Color::black:
                return false;
            }
        };
        
        for (auto start = 0; start != adj.size(); ++start) {
            if (has_cycle_from(start)) {
                out.clear();
                break;
            }
        }
        
        return out;
    }
}

vector<int> Solution::findOrder(const int numCourses,
                                const vector<pair<int, int>>& prerequisites)
{
    const auto adj = build_adjacency_list(numCourses, prerequisites);
    return reverse_topological_sort(adj);
}
